File: voca_crm_api/src/main/java/com/vocacrm/api/model/Reservation.java
package com.vocacrm.api.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.UUID;

/**
 * 예약 엔티티
 */
@Entity
@Table(name = "reservations")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Reservation {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "member_id", nullable = false)
    private UUID memberId;

    @Column(name = "business_place_id", nullable = false, length = 7)
    private String businessPlaceId;

    @Column(name = "reservation_date", nullable = false)
    private LocalDate reservationDate;

    @Column(name = "reservation_time", nullable = false)
    private LocalTime reservationTime;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "status", nullable = false, columnDefinition = "reservation_status")
    @Builder.Default
    private ReservationStatus status = ReservationStatus.PENDING;

    @Column(name = "service_type", length = 100)
    private String serviceType;

    @Column(name = "duration_minutes")
    @Builder.Default
    private Integer durationMinutes = 60;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "remark", length = 200)
    private String remark;

    @Column(name = "created_by")
    private UUID createdBy;

    @Column(name = "updated_by")
    private UUID updatedBy;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    // 연관관계 매핑 (지연 로딩)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", insertable = false, updatable = false)
    @ToString.Exclude
    @JsonIgnore
    private Member member;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "business_place_id", insertable = false, updatable = false)
    @ToString.Exclude
    @JsonIgnore
    private BusinessPlace businessPlace;

    /**
     * 예약 상태 ENUM
     */
    public enum ReservationStatus {
        PENDING("대기중"),
        CONFIRMED("확정됨"),
        CANCELLED("취소됨"),
        COMPLETED("완료됨"),
        NO_SHOW("노쇼");

        private final String displayName;

        ReservationStatus(String displayName) {
            this.displayName = displayName;
        }

        public String getDisplayName() {
            return displayName;
        }
    }

}

File: voca_crm/lib/domain/entity/reservation.dart
/// 예약 엔티티
class Reservation {
  final String id;
  final String memberId;
  final String businessPlaceId;
  final DateTime reservationDate;
  final DateTime reservationTime;
  final ReservationStatus status;
  final String? serviceType;
  final int durationMinutes;
  final String? notes;
  final String? remark; // 예약 특이사항 (예: "30분 늦을 수도 있음")
  final String? createdBy;
  final String? updatedBy; // 마지막 수정자
  final DateTime createdAt;
  final DateTime updatedAt;

  Reservation({
    required this.id,
    required this.memberId,
    required this.businessPlaceId,
    required this.reservationDate,
    required this.reservationTime,
    required this.status,
    this.serviceType,
    this.durationMinutes = 60,
    this.notes,
    this.remark,
    this.createdBy,
    this.updatedBy,
    required this.createdAt,
    required this.updatedAt,
  });

  Reservation copyWith({
    String? id,
    String? memberId,
    String? businessPlaceId,
    DateTime? reservationDate,
    DateTime? reservationTime,
    ReservationStatus? status,
    String? serviceType,
    int? durationMinutes,
    String? notes,
    String? remark,
    String? createdBy,
    String? updatedBy,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Reservation(
      id: id ?? this.id,
      memberId: memberId ?? this.memberId,
      businessPlaceId: businessPlaceId ?? this.businessPlaceId,
      reservationDate: reservationDate ?? this.reservationDate,
      reservationTime: reservationTime ?? this.reservationTime,
      status: status ?? this.status,
      serviceType: serviceType ?? this.serviceType,
      durationMinutes: durationMinutes ?? this.durationMinutes,
      notes: notes ?? this.notes,
      remark: remark ?? this.remark,
      createdBy: createdBy ?? this.createdBy,
      updatedBy: updatedBy ?? this.updatedBy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// 예약 상태 ENUM
enum ReservationStatus {
  PENDING,
  CONFIRMED,
  CANCELLED,
  COMPLETED,
  NO_SHOW;

  String get displayName {
    switch (this) {
      case ReservationStatus.PENDING:
        return '대기중';
      case ReservationStatus.CONFIRMED:
        return '확정됨';
      case ReservationStatus.CANCELLED:
        return '취소됨';
      case ReservationStatus.COMPLETED:
        return '완료됨';
      case ReservationStatus.NO_SHOW:
        return '노쇼';
    }
  }

  static ReservationStatus? fromString(String? value) {
    if (value == null) return null;
    try {
      return ReservationStatus.values.firstWhere(
        (e) => e.name.toUpperCase() == value.toUpperCase(),
      );
    } catch (e) {
      return null;
    }
  }
}
